equistore-torch/src/tensor.cpp

```diff
@@ -1,19 +1,432 @@
+            "TensorMap::keys_to_properties first argument must be a `str`, list of `str` or `Labels`"
+        );
+    }
+}
+
+TorchTensorMap TensorMapHolder::keys_to_samples(torch::IValue keys_to_move, bool sort_samples) const {
+    if (keys_to_move.isString() || keys_to_move.isList() || keys_to_move.isTuple()) {
+        auto selection = extract_list_str(keys_to_move, "TensorMap::keys_to_samples first argument");
+        auto tensor = tensor_.keys_to_samples(selection, sort_samples);
+        return torch::make_intrusive<TensorMapHolder>(std::move(tensor));
+    } else if (keys_to_move.isCustomClass()) {
+        auto selection = keys_to_move.toCustomClass<LabelsHolder>();
+        auto tensor = tensor_.keys_to_samples(selection->as_equistore(), sort_samples);
+        return torch::make_intrusive<TensorMapHolder>(std::move(tensor));
+    } else {
+        C10_THROW_ERROR(TypeError,
+            "TensorMap::keys_to_samples first argument must be a `str`, list of `str` or `Labels`"
+        );
+    }
+}
+
+TorchTensorMap TensorMapHolder::components_to_properties(torch::IValue dimensions) const {
+    auto selection = extract_list_str(dimensions, "TensorMap::components_to_properties argument");
+    auto tensor = this->tensor_.components_to_properties(selection);
+    return torch::make_intrusive<TensorMapHolder>(std::move(tensor));
 }
 
-TensorMapHolder::TensorMapHolder(
-    TorchLabels keys,
-    std::vector<TorchTensorBlock> blocks
-):
-    tensor_(keys->get(), blocks_from_torch(std::move(blocks)))
-{
-    throw std::runtime_error("not implemented");
+static std::vector<std::string> labels_names(const equistore::TensorBlock& block, size_t dimension) {
+    auto result = std::vector<std::string>();
+
+    auto labels = block.labels(dimension);
+    for (const auto& name: labels.names()) {
+        result.push_back(std::string(name));
+    }
+
+    return result;
+}
+
+std::vector<std::string> TensorMapHolder::sample_names() {
+    if (tensor_.keys().count() == 0) {
+        return {};
+    }
+
+    return labels_names(this->block_by_id(0)->as_equistore(), 0);
+}
+
+std::vector<std::vector<std::string>> TensorMapHolder::components_names() {
+    auto result = std::vector<std::vector<std::string>>();
+
+    if (tensor_.keys().count() != 0) {
+        auto block = this->block_by_id(0);
+        auto n_dimensions = block->values().sizes().size();
+
+        for (size_t dimension=1; dimension<n_dimensions-1; dimension++) {
+            result.push_back(labels_names(block->as_equistore(), dimension));
+        }
+    }
+
+    return result;
+}
+
+std::vector<std::string> TensorMapHolder::property_names() {
+    if (tensor_.keys().count() == 0) {
+        return {};
+    }
+
+    auto block = this->block_by_id(0);
+    auto n_dimensions = block->values().sizes().size();
+
+    return labels_names(block->as_equistore(), n_dimensions - 1);
+}
+
+std::vector<std::tuple<TorchLabelsEntry, TorchTensorBlock>> TensorMapHolder::items() {
```

Comment:
Since this is a simple wrapper, I would rather re-implement it on the Python side directly. This reduces the number of functions/types in the shared library, reducing the cost to add new wrappers